crypto/objects/obj_dat.c

```diff
@@ -37,6 +38,43 @@ struct added_obj_st {
 
 static int new_nid = NUM_NID;
 static LHASH_OF(ADDED_OBJ) *added = NULL;
+static CRYPTO_RWLOCK *ossl_obj_lock = NULL;
+
+static int ossl_obj_ln2nid(const char *s, int lock);
+static int ossl_obj_sn2nid(const char *s, int lock);
+
+static CRYPTO_ONCE ossl_obj_lock_init = CRYPTO_ONCE_STATIC_INIT;
+
+DEFINE_RUN_ONCE_STATIC(obj_lock_initialise)
+{
+    ossl_obj_lock = CRYPTO_THREAD_lock_new();
+    return ossl_obj_lock != NULL;
+}
+
+static int ossl_init_added_lock(void)
+{
+    return RUN_ONCE(&ossl_obj_lock_init, obj_lock_initialise);
+}
+
+static int ossl_obj_write_lock(int lock)
+{
+    if (!ossl_init_added_lock())
+        return 0;
+    return !lock || CRYPTO_THREAD_write_lock(ossl_obj_lock);
+}
+
+static int ossl_obj_read_lock(int lock)
+{
+    if (!ossl_init_added_lock())
+        return 0;
+    return !lock || CRYPTO_THREAD_read_lock(ossl_obj_lock);
+}
+
+static void ossl_obj_unlock(int lock)
```

Comment:
Could the above 4 static functions be inline? Or macros?

Reply:
They could, I'm assuming that the compiler will inline them for us without being told.
This is true of most compilers.

Reply:
Yeah, it is, but it doesn't hurt to provide a hint.

Reply:
Code bloat and premature optimisation?